File: gtci/12-tree-depth-first-search/python3/3-all-paths-for-a-sum.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findPaths(self, root, required_sum):
        results = []

        self.findPaths_helper(results, root, required_sum)

        return results

    def findPaths_helper(self, results, root, required_sum, path=[]):
        if root is None:
            return

        path.append(root.val)

        if root.left is None and root.right is None and root.val == required_sum:
            results.append(list(path))
        else:
            self.findPaths_helper(results, root.left, required_sum - root.val, path)
            self.findPaths_helper(results, root.right, required_sum - root.val, path)

        path.pop()  # backtrack


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.left.right = TreeNode(-3)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(3)

    assert sol.findPaths(root, 23) == [[12, 1, 10]]
    assert sol.findPaths(root, 16) == [[12, 7, -3], [12, 1, 3]]

    print("All test cases passed.")

File: gtci/12-tree-depth-first-search/cpp/3-all-paths-for-a-sum.cpp
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

#include "tree_utils.h"

class Solution {
public:
  vector<vector<int>> findPaths(TreeNode* root, int sum) {
    vector<vector<int>> allPaths;
    vector<int> path;
    
    findPathsHelper(root, sum, allPaths, path);

    return allPaths;
  }

  void findPathsHelper(TreeNode* root, int sum, std::vector<std::vector<int>>& results, std::vector<int>& path) {
    if (root == nullptr)
      return;

    path.push_back(root->val);
    
    if (root->left == nullptr && root->right == nullptr && root->val == sum) {
      results.push_back(path);
    } else {
      findPathsHelper(root->left, sum - root->val, results, path);
      findPathsHelper(root->right, sum - root->val, results, path);
    }

    path.pop_back();
  }
};

int main() {
    Solution sol;
    TreeNode* root = new TreeNode(12);
    root->left = new TreeNode(7);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(9);
    root->left->right = new TreeNode(-3);
    root->right->left = new TreeNode(10);
    root->right->right = new TreeNode(3);

    assert(sol.findPaths(root, 23) == vector<vector<int>>({{12, 1, 10}}));
    assert(sol.findPaths(root, 16) == vector<vector<int>>({{12, 7, -3}, {12, 1, 3}}));

    delete root;

    std::cout << "All test cases passed.\n";

    return 0;
}
